### DISCUSSION TITLE
Indirect modification of overloaded element of Illuminate\Database\Eloquent\Collection has no effect

### DISCUSSION THREAD

**Discussion by pedro-saraiva88** [livewire/livewire#4160] [1. Help] [answered]
I'am getting this error **_indirect modification of overloaded element of Illuminate\Database\Eloquent\Collection has no effect_**  everytime i type on a child related model after adding a new city (related model).

```php

namespace App\Http\Livewire;

use App\Models\City;
use App\Models\Country as ModelsCountry;
use Livewire\Component;

class Country extends Component
{
    public ModelsCountry $country;

    protected $rules = [
        'country.name' => 'required|string|max:255|min:4',
        'country.cities'=>'min:3',
        'country.cities.*.name'=>"required"
        
    ];

    public function addCity()
    {
        $this->country->cities->push(new City);        
    }

    public function save()
    {
        $this->validate();
        $this->country->save();
    }

    public function mount(ModelsCountry $country)
    {
        $this->country = $country;
        
    }
    
    public function render()
    {
        return view('livewire.country');
    }
}
```

```php 

<div class="">

    <x-ui::form.input-row name="country.name" label="Name" wire:model="country.name" />

    @foreach ($country->cities as $index => $city)
        <x-ui::form.input-row 
            name="country.cities.{{$index}}.name" 
            label="Name" wire:model="country.cities.{{$index}}.name" 
            wire:key="country.cities.{{$index}}.name"
        />

        
    @endforeach
        
    <x-ui::button.info wire:click="addCity" class="mt-4">Add City</x-ui::button.info>
    <x-ui::button.success wire:click="save" class="mt-4">Save</x-ui::button.success>
</div>


```

**Comment by godismyjudge95** [comment, marked as answer]
I believe this is because of this line here:
```
$this->country->cities->push(new City);        
```

You are pushing a city onto the cities collection instead of saving it to the relation.  To save a new city to the country you would do the following instead:
```
$this->country->cities()->save(new City);
```

Like the error says, you were modifying the collection instead of saving to the relationship.  Modifying an eloquent collection doesn't allow you to save those changes to the database.

**Reply by pedro-saraiva88** [reply]
Many thanks for the reply godismyjudge95.
I am not saving the City to the database because I would like to be able to add multiple cities and only save them to the database when I click the submit button.
Another problem is that if I'm adding a new country (parent relationship), it only allows me to save the cities after entering the parent in the database.
What is the best way to do this?

**Reply by godismyjudge95** [reply]
Ah ok, if you are wanting to add a city without saving it to the database you can do the following:
```
$this->country->cities()->associate(new City);
```

Then later you can call:
```
$this->country->save();
```

**Reply by pedro-saraiva88** [reply]
I changed the addCity function to this

```php

	public function addCity()
	{
		$this->validate([
			'new_city' => 'required|max:255'
		]);

		$city = new City;
		$city->name = $this->new_city;

		$this->country->cities()->associate($city);
		$this->new_city = '';
	}

```

Now when i click to add a new city i'm getting this error _**BadMethodCallException
Call to undefined method Illuminate\Database\Eloquent\Relations\HasMany::attach()**_

Note: I can't find any good example of the new feature **Deep model data binding** 😩

**Reply by godismyjudge95** [reply]
Ah I forgot that cities is a HasMany relation on your Country model.  I think the only way to do this then, would be to keep an array of cities in your component:
```
public array $citites = [];
```

And then associate the city with the country and save it to the cities array:
```
public function addCity()
{
	$this->validate([
		'new_city' => 'required|max:255'
	]);

	$city = new City;
	$city->name = $this->new_city;

	$city->country()->associate($this->country);
	$this->new_city = '';

        $this->cities[] = $city;
}
```

And finally when you need to save the cities you would do:
```
public function save()
{
        foreach ($this->cities as $city) {
                $city->save();
        }
}
```

You might need to change your loop to use the new cities array too:
```
    @foreach ($cities as $index => $city)
        <x-ui::form.input-row 
            name="cities.{{$index}}.name" 
            label="Name" wire:model="cities.{{$index}}.name" 
            wire:key="cities.{{$index}}.name"
        />
    @endforeach
```

Hope this works for you.

**Comment by pedro-saraiva88** [comment]
Thank you godismyjudge95, i will try your suggestion.
